### PR TITLE
Make DNS spoofing compatible with kube-dns and different setups

### LINKED ISSUE #390: User Issue: Node level DNS disruptions impact the DNS records of all cluster's Kubernetes Services
**Describe the bug**
While testing the DNS disruptions at node level I noticed a couple of critical issues.

In summary, these impact the DNS records of all cluster's Kubernetes Services (`*.svc.cluster.local`).

**To Reproduce**
1. Start the provided minikube setup
    - `make minikube-start`
    - `make minikube-build`
    - `make install`
2. Do a `nslookup` before applying the Disruption:
    ```
    / # cat /etc/resolv.conf
    search chaos-demo.svc.cluster.local svc.cluster.local cluster.local
    nameserver 10.96.0.10
    options ndots:5
    
    Name:      demo.chaos-demo.svc.cluster.local
    Address 1: 10.108.230.236 demo.chaos-demo.svc.cluster.local
    
    Name:      dashboard-metrics-scraper.kubernetes-dashboard.svc.cluster.local
    Address 1: 10.107.89.68 dashboard-metrics-scraper.kubernetes-dashboard.svc.cluster.local
    ```
3. Curling the 2 Services from the curl pod works as expected. The same applies to any external hostnames like `google.com`.
4. Apply the following Custom Resource:
    ```
    apiVersion: chaos.datadoghq.com/v1beta1
    kind: Disruption
    metadata:
      name: dns
      namespace: chaos-engineering
    spec:
      level: node
      selector:
        kubernetes.io/hostname: minikube
      count: 100%
      dns:
        - hostname: demo.chaos-demo.svc.cluster.local
          record:
            type: A
            value: 10.0.0.154,10.0.0.13  
    ```
5. Do a `nslookup` again:
    ```
    / # nslookup demo.chaos-demo.svc.cluster.local
    Name:      demo.chaos-demo.svc.cluster.local
    Address 1: 10.0.0.154
    
    / # nslookup dashboard-metrics-scraper.kubernetes-dashboard.svc.cluster.local
    nslookup: can't resolve 'dashboard-metrics-scraper.kubernetes-dashboard.svc.cluster.local': Name does not resolve
    ```
6. Curling the 2 Services from the curl pod is no more working. External hostnames are still accessible.

**Expected behavior**
The disruption should only impact the provided hostname (`demo.chaos-demo.svc.cluster.local`).

**Environment:**
 - Kubernetes version: 1.21.1 in the provided minikube setup
 - Controller version:
 - Cloud provider (or `local`): minikube
 - Base OS for Kubernetes: N/A

**Additional context**
In minikube there is only one node which means all outgoing calls to the cluster's Kubernetes Services are affected. In a multi-node setup this affects the nodes which are targeted using the label selector.

### PR DESCRIPTION
**Pull request by nikos912000**
## What does this PR do?

- [ ] Adds new functionality
- [ ] Alters existing functionality
- [x] Fixes a bug
- [x] Improves Documentation

Fixes #390.

In the current implementation of the DNS spoofing unmatched requests (i.e. the ones which are not aimed for the attacked hostname) are forwarded to Google's Public DNS server, 8.8.8.8. As explained in #390 this has a couple of issues:
- When CoreDNS/kube-dns is used we should forward the requests to that instead. However, this depends on the setup:
  - In Minikube, forwarding internal requests to kube-dns and external ones to a public DNS works.
  - In EKS, all requests can be forwarded to kube-dns (see [here](https://aws.amazon.com/premiumsupport/knowledge-center/eks-dns-failure/)).
- There is no way to override the upstream DNS server as this value is hardcoded and not exposed as a parameter from the injector/controller.

This PR introduces a backwards-compatible way to support different setups for DNS spoofing. It relies on 2 flags which can be passed to the controller:
- `dnsDisruption.dnsServer`: this was hardcoded to 8.8.8.8 before while now it is configurable
- `dnsDisruption.kubeDns`: this can be one of (`off`, `internal`, `all`). See the updated docs for more context.

**Useful links**
- https://pracucci.com/kubernetes-dns-resolution-ndots-options-and-why-it-may-affect-application-performances.html
- https://aws.amazon.com/premiumsupport/knowledge-center/eks-dns-failure/

### Testing

- [x] I used existing unit tests
- [x] I manually tested locally
- [x] I will manually test in a canary deployment

Please list your manual testing steps (sample `yaml` file, commands, important checks):

### Checklist

- [x] The documentation is up to date
- [x] My code is sufficiently commented
- [x] I have tested to the best of my ability
- [x] I have signed my commit (see [Contributing Docs](../CONTRIBUTING.md))

### CODE REVIEW

File: api/disruption_kind.go [hunk 1]
```diff
@@ -45,6 +45,10 @@ func AppendArgs(args []string, level chaostypes.DisruptionLevel, kind chaostypes
 		args = append(args, "--on-init")
 	}
 
+	// DNS disruption configs
+	args = append(args, "--dns-server", dnsServer)
+	args = append(args, "--kube-dns", kubeDns)
```
Comment (ptnapoleon) [inline api/disruption_kind.go:50]: I think these could be gated under an `if kind == chaostypes.DisruptionKindDNSDisruption`, yes?

**Review by ptnapoleon — COMMENTED** [review]
Only one request, along with making the linter happy, but otherwise this is very great, thanks!

File: api/disruption_kind.go (hunk 1 above)
Comment (nikos912000) [reply api/disruption_kind.go]: Yep, I missed that, thanks.

Comment (ptnapoleon) [reply api/disruption_kind.go]: looks like this was added and then removed again?

Comment (nikos912000) [reply api/disruption_kind.go]: Indeed. I've pushed the change again.

**Comment by nikos912000** [comment]
Btw, I did a test with the latest changes but something looks wrong:
```
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
main.initConfig()
/Users/nkatirtzis/github/chaos-controller/cli/injector/main.go:206 +0x16df
```

Is it possible that [5b346e1db31203134a951ba38b7c45a24938bd0c](https://github.com/DataDog/chaos-controller/commit/5b346e1db31203134a951ba38b7c45a24938bd0c) introduced breaking changes?

**Comment by ptnapoleon** [comment]
> Btw, I did a test with the latest changes but something looks wrong:
> 
> ```
> panic: runtime error: index out of range [0] with length 0
> 
> goroutine 1 [running]:
> main.initConfig()
> /Users/nkatirtzis/github/chaos-controller/cli/injector/main.go:206 +0x16df
> ```
> 
> Is it possible that [5b346e1db31203134a951ba38b7c45a24938bd0c](https://github.com/DataDog/chaos-controller/commit/5b346e1db31203134a951ba38b7c45a24938bd0c) introduced breaking changes?

Well it's definitely possible. Are you just running `make test`? What's your go version?

**Comment by nikos912000** [comment]
> Well it's definitely possible. Are you just running `make test`? What's your go version?
No, I deployed to minikube following the [instructions](https://github.com/DataDog/chaos-controller/blob/main/CONTRIBUTING.md#deploying-local-changes-to-minikube) so hopefully this can be reproduced.

Also, I noticed that after updating the Makefile to use `image load` instead of ssh-ing the images are always cached which might result to misleading tests.

**Review by ptnapoleon — DISMISSED** [review]

**Comment by ptnapoleon** [comment]
> Also, I noticed that after updating the Makefile to use `image load` instead of ssh-ing the images are always cached which might result to misleading tests.

Oh hmm, that is really good to know. I always destroy and re-create my minikube cluster, so I've avoided that, but I'm sure it's affecting others

**Comment by ptnapoleon** [comment]
I have no issues running an injection, DNS or otherwise. I am a little surprised that my injector pods have "--kube-dns 0" and not "off"?

```
    Args:
      dns-disruption
      --host-record-pairs
      consul.*;A;8.8.8.8
      --host-record-pairs
      foo.bar.svc.cluster.local;A;10.0.0.154,10.0.0.13,10.0.0.44
      --host-record-pairs
      datadoghq.com;CNAME;google.com
      --metrics-sink
      noop
      --level
      pod
      --containers-id
      containerd://b2001335d2130ce767be43dbcf74f811c779db24cbec1e45457d62456ae56126
      --log-context-disruption-name
      disruption-sample
      --log-context-disruption-namespace
      chaos-engineering
      --log-context-target-name
      hello-node-7567d9fdc9-p6mlx
      --dns-server
      8.8.8.8
      --kube-dns
      0
```

**Comment by nikos912000** [comment]
> I have no issues running an injection, DNS or otherwise. I am a little surprised that my injector pods have "--kube-dns 0" and not "off"?
> 
> ```
>     Args:
>       dns-disruption
>       --host-record-pairs
>       consul.*;A;8.8.8.8
>       --host-record-pairs
>       foo.bar.svc.cluster.local;A;10.0.0.154,10.0.0.13,10.0.0.44
>       --host-record-pairs
>       datadoghq.com;CNAME;google.com
>       --metrics-sink
>       noop
>       --level
>       pod
>       --containers-id
>       containerd://b2001335d2130ce767be43dbcf74f811c779db24cbec1e45457d62456ae56126
>       --log-context-disruption-name
>       disruption-sample
>       --log-context-disruption-namespace
>       chaos-engineering
>       --log-context-target-name
>       hello-node-7567d9fdc9-p6mlx
>       --dns-server
>       8.8.8.8
>       --kube-dns
>       0
> ```

Cool, let me go through that and do some testing.

I had that as a boolean initially but I changed it to a string. I'll check if all my changes were pushed.

**Comment by ptnapoleon** [comment]
It's "off" in the config map, and is 0 as early as the controller's main. go, I'm trying to see if I'm missing anything about how we load the arg from the config map

**Comment by nikos912000** [comment]
@ptnapoleon I've given this another attempt on the main branch, without my changes, but I was getting the error message mentioned above.
I raised an [issue](https://github.com/DataDog/chaos-controller/issues/409) on how to reproduce if it helps.

**Comment by nikos912000** [comment]
I did another test after merging the fix from master and updating the ConfigMap and it's looking good.
Feel free to merge if there is nothing else on your side.

**Comment by nikos912000** [comment]
@ptnapoleon I know you released version 4.4.1 right before that PR got merged, but would it be possible to do another one for this?
Maybe once https://github.com/DataDog/chaos-controller/pull/400 gets merged as well as these 2 are pretty useful for us.

**Comment by ptnapoleon** [comment]
We'll release right after #400 is merged, which should be Monday

**Pull request merged**